File: code_just_in_case/data_collector.js
import fs from 'fs';
import snoowrap from 'snoowrap';
import { createClient } from '@supabase/supabase-js';
import * as tf from '@tensorflow/tfjs-node';
import { UniversalSentenceEncoder } from '@tensorflow-models/universal-sentence-encoder';

// Import configuration - you'll need to create a config.js file
import {
    REDDIT_CLIENT_ID,
    REDDIT_CLIENT_SECRET,
    REDDIT_USER_AGENT,
    SUPABASE_URL,
    SUPABASE_ANON_KEY,
} from './config.js';

// Supabase setup
const supabase = createClient(SUPABASE_URL, SUPABASE_ANON_KEY);

// Reddit API setup
const reddit = new snoowrap({
    userAgent: REDDIT_USER_AGENT,
    clientId: REDDIT_CLIENT_ID,
    clientSecret: REDDIT_CLIENT_SECRET,
    refreshToken: undefined, // For read-only access
    accessToken: undefined,
});

// Universal Sentence Encoder model (will be loaded asynchronously)
let model = null;

/**
 * Load the Universal Sentence Encoder model
 */
async function loadModel() {
    if (!model) {
        console.log('Loading Universal Sentence Encoder model...');
        model = await UniversalSentenceEncoder.load();
        console.log('Model loaded successfully!');
    }
    return model;
}

/**
 * Calculate cosine similarity between two vectors
 */
function cosineSimilarity(vecA, vecB) {
    const dotProduct = tf.sum(tf.mul(vecA, vecB));
    const normA = tf.norm(vecA);
    const normB = tf.norm(vecB);
    return tf.div(dotProduct, tf.mul(normA, normB));
}

/**
 * Get the most recent submission from Supabase
 */
async function getLatestSubmission() {
    try {
        const { data, error } = await supabase
            .from('submissions')
            .select('subreddit, submission_flair, is_nsfw, title, content')
            .order('id', { ascending: false })
            .limit(1);

        if (error) {
            console.error('Supabase error:', error);
            return null;
        }

        return data && data.length > 0 ? data[0] : null;
    } catch (error) {
        console.error('Error fetching latest submission:', error);
        return null;
    }
}

/**
 * Test if Reddit API credentials are working
 */
async function testRedditConnection() {
    try {
        const subreddit = reddit.getSubreddit('python');
        await subreddit.getHot({ limit: 1 });
        console.log('Reddit connection test: Connection succeeded.');
        return true;
    } catch (error) {
        console.error('Reddit connection failed:', error);
        return false;
    }
}

/**
 * Fetch posts from Reddit with filtering
 */
async function fetchRedditPosts(subredditName, submissionFlair = null, isNsfw = false, limit = 100) {
    const cleanSubredditName = subredditName.replace('r/', '');
    
    try {
        const subreddit = reddit.getSubreddit(cleanSubredditName);
        const submissions = await subreddit.getHot({ limit: limit });
        
        const posts = [];
        
        for (const submission of submissions) {
            // Filter NSFW
            if (submission.over_18 !== isNsfw) {
                continue;
            }
            
            // Filter flair if specified
            if (submissionFlair && submission.link_flair_text !== submissionFlair) {
                continue;
            }
            
            posts.push({
                title: submission.title,
                content: submission.selftext || submission.title,
                url: submission.url,
                score: submission.score,
                flair: submission.link_flair_text,
                nsfw: submission.over_18,
                id: submission.id,
            });
        }
        
        return posts;
    } catch (error) {
        console.error('Error fetching Reddit posts:', error);
        return [];
    }
}

/**
 * Find most similar posts using Universal Sentence Encoder embeddings + cosine similarity
 */
async function findSimilarPostsEmbeddings(targetPost, redditPosts, topK = 10) {
    if (!redditPosts || redditPosts.length === 0) {
        return [];
    }
    
    const encoder = await loadModel();
    
    const targetText = `${targetPost.title} ${targetPost.content}`;
    const redditTexts = redditPosts.map(post => `${post.title} ${post.content}`);
    
    // Encode with Universal Sentence Encoder
    const targetEmbedding = await encoder.embed([targetText]);
    const redditEmbeddings = await encoder.embed(redditTexts);
    
    // Calculate similarities
    const similarities = [];
    const targetVec = targetEmbedding.arraySync()[0];
    const redditVecs = redditEmbeddings.arraySync();
    
    for (let i = 0; i < redditVecs.length; i++) {
        const targetTensor = tf.tensor1d(targetVec);
        const redditTensor = tf.tensor1d(redditVecs[i]);
        const similarity = await cosineSimilarity(targetTensor, redditTensor);
        similarities.push(await similarity.data());
        
        // Clean up tensors
        targetTensor.dispose();
        redditTensor.dispose();
    }
    
    // Clean up embeddings
    targetEmbedding.dispose();
    redditEmbeddings.dispose();
    
    // Get top k indices
    const indexedSimilarities = similarities.map((sim, idx) => ({ sim: sim[0], idx }));
    indexedSimilarities.sort((a, b) => b.sim - a.sim);
    const topIndices = indexedSimilarities.slice(0, topK);
    
    const similarPosts = [];
    for (const { sim, idx } of topIndices) {
        const post = { ...redditPosts[idx] };
        post.similarity_score = sim;
        similarPosts.push(post);
    }
    
    return similarPosts;
}

/**
 * Main function to collect and structure all the data
 */
async function collectData() {
    // Test Reddit connection
    if (!(await testRedditConnection())) {
        console.log('Please check your Reddit API credentials');
        return null;
    }
    
    // Get latest submission
    const latestSubmission = await getLatestSubmission();
    if (!latestSubmission) {
        console.log('No submissions found in database');
        return null;
    }
    
    console.log(`Latest submission: ${latestSubmission.title}`);
    console.log(`Subreddit: ${latestSubmission.subreddit}`);
    console.log(`Flair: ${latestSubmission.submission_flair}`);
    console.log(`NSFW: ${latestSubmission.is_nsfw}`);
    console.log('-'.repeat(50));
    
    // Fetch Reddit posts
    const redditPosts = await fetchRedditPosts(
        latestSubmission.subreddit,
        latestSubmission.submission_flair,
        latestSubmission.is_nsfw,
        200
    );
    
    console.log(`Found ${redditPosts.length} matching Reddit posts`);
    
    if (redditPosts.length === 0) {
        console.log('No matching posts found on Reddit');
        return null;
    }
    
    // Find similar posts
    const similarPosts = await findSimilarPostsEmbeddings(latestSubmission, redditPosts, 10);
    
    console.log('\nTop 10 most similar posts (using Universal Sentence Encoder):');
    console.log('='.repeat(50));
    
    const finalData = [];
    
    for (let i = 0; i < similarPosts.length; i++) {
        const post = similarPosts[i];
        console.log(`${i + 1}. Title: ${post.title}...`);
        console.log(`   Score: ${post.score} | Similarity: ${post.similarity_score.toFixed(3)}`);
        console.log(`   Flair: ${post.flair} | NSFW: ${post.nsfw}`);
        console.log(`   URL: ${post.url}`);
        console.log('-'.repeat(40));
        
        const postData = { ...post };
        postData.top_level_comments = [];
        
        try {
            const submission = reddit.getSubmission(post.id);
            await submission.expandReplies({ limit: 10, depth: 1 });
            
            const comments = submission.comments.slice(0, 10);
            
            for (const comment of comments) {
                if (postData.top_level_comments.length >= 10) {
                    break;
                }
                
                if (comment.body && comment.body !== '[deleted]' && comment.body !== '[removed]') {
                    const commentData = {
                        score: comment.score,
                        body: comment.body,
                        author: comment.author ? comment.author.name : '[deleted]',
                        author_hot_comments: []
                    };
                    
                    if (comment.author && comment.author.name !== '[deleted]') {
                        try {
                            const authorComments = await comment.author.getComments({ sort: 'hot', limit: 10 });
                            
                            for (const authComment of authorComments) {
                                if (authComment.body && authComment.body !== '[deleted]' && authComment.body !== '[removed]') {
                                    commentData.author_hot_comments.push({
                                        score: authComment.score,
                                        body: authComment.body
                                    });
                                }
                            }
                        } catch (error) {
                            console.log(`   Error fetching comments for ${comment.author.name}: ${error.message}`);
                        }
                    }
                    
                    postData.top_level_comments.push(commentData);
                }
            }
        } catch (error) {
            console.log(`   Error fetching comments for post ${post.id}: ${error.message}`);
        }
        
        finalData.push(postData);
    }
    
    console.log('\nAll data has been saved to the finalData array.');
    return finalData;
}

// Main execution
async function main() {
    try {
        const finalData = await collectData();
        if (finalData) {
            fs.writeFileSync('reddit_data.json', JSON.stringify(finalData, null, 2));
            console.log('Data successfully written to reddit_data.json');
        }
    } catch (error) {
        console.error('Error in main execution:', error);
    } finally {
        // Clean up TensorFlow resources
        if (model) {
            model.dispose();
        }
    }
}

// Run if this file is executed directly
if (import.meta.url === `file://${process.argv[1]}`) {
    main();
}

export { collectData, getLatestSubmission, fetchRedditPosts, findSimilarPostsEmbeddings };

File: api/data_collector.py
import os
import json
import praw
import numpy as np
from supabase import create_client, Client
from sklearn.metrics.pairwise import cosine_similarity
from sentence_transformers import SentenceTransformer
from api.config import (
    REDDIT_CLIENT_ID,
    REDDIT_CLIENT_SECRET,
    REDDIT_USER_AGENT,
    SUPABASE_URL,
    SUPABASE_ANON_KEY,
)

# Supabase setup
supabase: Client = create_client(SUPABASE_URL, SUPABASE_ANON_KEY)

# Reddit API setup
reddit = praw.Reddit(
    client_id=REDDIT_CLIENT_ID,
    client_secret=REDDIT_CLIENT_SECRET,
    user_agent=REDDIT_USER_AGENT,
)

# Sentence Transformer model
model = SentenceTransformer("sentence-transformers/all-MiniLM-L6-v2")


def get_latest_submission():
    """Get the most recent submission from Supabase"""
    response = (
        supabase.table("submissions")
        .select("subreddit, submission_flair, is_nsfw, title, content")
        .order("id", desc=True)
        .limit(1)
        .execute()
    )

    if response.data:
        return response.data[0]
    return None


def test_reddit_connection():
    """Test if Reddit API credentials are working"""
    try:
        subreddit = reddit.subreddit("python")
        print(f"Reddit connection test: Connection succeeded.")
        return True
    except Exception as e:
        print(f"Reddit connection failed: {e}")
        return False


def fetch_reddit_posts(subreddit_name, submission_flair=None, is_nsfw=False, limit=100):
    """Fetch posts from Reddit with filtering"""
    subreddit_name = subreddit_name.replace("r/", "")

    try:
        subreddit = reddit.subreddit(subreddit_name)
        posts = []

        for submission in subreddit.hot(limit=limit):
            # Filter NSFW
            if submission.over_18 != is_nsfw:
                continue

            # Filter flair if specified
            if submission_flair and submission.link_flair_text != submission_flair:
                continue

            posts.append(
                {
                    "title": submission.title,
                    "content": submission.selftext if submission.selftext else submission.title,
                    "url": submission.url,
                    "score": submission.score,
                    "flair": submission.link_flair_text,
                    "nsfw": submission.over_18,
                    "id": submission.id,
                }
            )

        return posts
    except Exception as e:
        print(f"Error fetching Reddit posts: {e}")
        return []


def find_similar_posts_embeddings(target_post, reddit_posts, top_k=10):
    """Find most similar posts using SentenceTransformers embeddings + cosine similarity"""
    if not reddit_posts:
        return []

    target_text = f"{target_post['title']} {target_post['content']}"
    reddit_texts = [f"{post['title']} {post['content']}" for post in reddit_posts]

    # Encode with sentence transformer
    target_embedding = model.encode([target_text], convert_to_numpy=True, normalize_embeddings=True)
    reddit_embeddings = model.encode(reddit_texts, convert_to_numpy=True, normalize_embeddings=True)

    similarities = cosine_similarity(target_embedding, reddit_embeddings)[0]

    # Get top k
    top_indices = np.argsort(similarities)[::-1][:top_k]

    similar_posts = []
    for idx in top_indices:
        post = reddit_posts[idx].copy()
        post["similarity_score"] = float(similarities[idx])
        similar_posts.append(post)

    return similar_posts


def collect_data():
    """Main function to collect and structure all the data."""
    if not test_reddit_connection():
        print("Please check your Reddit API credentials")
        return None

    latest_submission = get_latest_submission()
    if not latest_submission:
        print("No submissions found in database")
        return None

    print(f"Latest submission: {latest_submission['title']}")
    print(f"Subreddit: {latest_submission['subreddit']}")
    print(f"Flair: {latest_submission['submission_flair']}")
    print(f"NSFW: {latest_submission['is_nsfw']}")
    print("-" * 50)

    reddit_posts = fetch_reddit_posts(
        subreddit_name=latest_submission["subreddit"],
        submission_flair=latest_submission["submission_flair"],
        is_nsfw=latest_submission["is_nsfw"],
        limit=200,
    )

    print(f"Found {len(reddit_posts)} matching Reddit posts")

    if not reddit_posts:
        print("No matching posts found on Reddit")
        return None

    similar_posts = find_similar_posts_embeddings(latest_submission, reddit_posts, top_k=10)

    print("\nTop 10 most similar posts (using Sentence Transformers):")
    print("=" * 50)
    
    final_data = []

    for i, post in enumerate(similar_posts, 1):
        print(f"{i}. Title: {post['title']}...")
        print(f"   Score: {post['score']} | Similarity: {post['similarity_score']:.3f}")
        print(f"   Flair: {post['flair']} | NSFW: {post['nsfw']}")
        print(f"   URL: {post['url']}")
        print("-" * 40)
        
        post_data = post.copy()
        post_data['top_level_comments'] = []

        try:
            submission = reddit.submission(id=post['id'])
            submission.comment_sort = 'top'

            for comment in submission.comments:
                if len(post_data['top_level_comments']) >= 10:
                    break
                
                if not isinstance(comment, praw.models.MoreComments) and comment.body:
                    comment_data = {
                        "score": comment.score,
                        "body": comment.body,
                        "author": comment.author.name if comment.author else "[deleted]",
                        "author_hot_comments": []
                    }
                    
                    if comment.author and comment.author.name != "[deleted]":
                        try:
                            for auth_comment in comment.author.comments.hot(limit=10):
                                if not isinstance(auth_comment, praw.models.MoreComments) and auth_comment.body:
                                    comment_data['author_hot_comments'].append({
                                        "score": auth_comment.score,
                                        "body": auth_comment.body
                                    })
                        except Exception as e:
                            print(f"   Error fetching comments for {comment.author.name}: {e}")
                            
                    post_data['top_level_comments'].append(comment_data)
        
        except Exception as e:
            print(f"   Error fetching comments for post {post['id']}: {e}")
        
        final_data.append(post_data)

    print("\nAll data has been saved to the 'final_data' list.")
    return final_data

if __name__ == "__main__":
    final_data = collect_data()
    if final_data:
        with open("reddit_data.json", "w") as f:
            json.dump(final_data, f, indent=2)
        print("Data successfully written to reddit_data.json")